src/pandaboxlib/pandaboxlib.py

```diff
@@ -6,80 +6,282 @@
+        The blocks referenced in a design must be available in the currently
+        installed FPGA app. This can be ensured in a semi-automatic fashion by
+        specifying the required FW version in the saved design and subsequently
+        validating it when loading a design (see ``_validate_fw_version``).
+
+        N.b. Whilst the entire system identification string is saved, only the
+        FPGA field (i.e. installed FPGA app) is likely of interest.
+
+        """
+        idn = self._read_idn()
+        file.write(f"*ECHO {idn}?\n")      # Save as ECHO to avoid attempted assignment on load
+
+    def _validate_fw_version(self, file):
+        """Validate firmware version
+
+        The blocks referenced in a design must be available in the currently
+        installed FPGA app. This can be ensured in a semi-automatic fashion by
+        specifying the required FW version in the saved design (see
+        ``_save_fw_version``) and subsequently validating it when loading a
+        design.
+
+        N.b. Whilst the entire system identification string is saved, only the
+        FPGA field (i.e. installed FPGA app) is likely of interest.
+
+        """
+
+        def parse_fpga_from_idn(idn):
+            return idn.split(": ")[2]
+
+        # Read installed FPGA FW version
+        fpga_installed = parse_fpga_from_idn(self._read_idn())
+
+        # Read design FPGA FW version
+        offset = file.tell()
+        file.seek(0)
+        fpga_design = parse_fpga_from_idn(file.readline())
+        file.seek(offset)
+
+        # Validate
+        return fpga_design == fpga_installed
+
+    def save(self, path):
+        """Save design to file"""
+        with open(path, "wt") as file:
+
+            # Save firmware version
+            self._save_fw_version(file)
+
+            # Save the CONFIG state
+            self._save_state(file, '*CHANGES.ATTR?')
+            self._save_state(file, '*CHANGES.CONFIG?')
+
+            # Save individual tables
+            #
+            #   N.b. Must read complete table response before processing!
+            #
+            tables = list(self._read_response('*CHANGES.TABLE?'))
+            for table in tables:
+                self._save_table(file, table)
+
+            # Save metadata
+            #
+            #   N.b. Table and value results are mixed together
+            #
+            metadata = list(self._read_response('*CHANGES.METADATA?'))
+            for line in metadata:
+                self._save_metadata(file, line)
+
+    def load(self, path):
+        """Load design from file"""
+        with open(path,"rt") as file:
+
+            # Validate FPGA firmware version
+            if not self._validate_fw_version(file):
+                raise ValueError("Design and installed FPGA FW versions do not match")
+
+            # Apply design
+            for line in file:
+                self.sock.sendall(line.encode())
 
 
 class PandA:
 
-    def __init__(self, host):
+    sock = None
```

Comment:
is this required?

Reply:
Previously, the socket was instantiated in `PandA.__init__`, connected in `PandA.connect_to_panda` and closed in to `PandA.disconnect_from_panda`. As such, a call to `PandA.disconnect_from_panda` [rendered the socket instance useless](https://docs.python.org/3.7/library/socket.html?highlight=socket#socket.socket.close), and a new instance could only be created by creating a new `PandA` instance.

To avoid this and allow connections to the pandabox to be opened and closed multiple times, the socket is now instantiated and connected in `PandA.connect_to_panda`. `PandA.sock` is then set to `None` after closing the socket to show that no active socket instance exists.

As such, I think it needs to be explicitly set to `None` here (or in `PandA.__init__`) to ensure that later references to it don't fail (see L227: `PandA.connect_to_panda`, (L222: `PandA.__del__`, L239: `PandA.disconnect_from_panda`).

Reply:
isn't this be `self.sock = None` then?

Reply:
I don't think so; `self` isn't defined in this scope.
Here `sock` is a class attribute rather than an instance attribute:
```python
>>> class Foo:
...   bar = 6.2831
... 
>>> Foo.bar
6.2831
>>> foo = Foo()
>>> foo.bar
6.2831
>>> Foo.bar = 3.1415
>>> foo.bar
3.1415
```

Reply:
thanks for the explanation, didn't know about class attributes.